### DISCUSSION TITLE
brew on symlinked home

### DISCUSSION THREAD

**Discussion by mrckndt** [Homebrew/discussions#1282] [Linux]
Running brew on Fedora Silverblue is a bit odd because one needs to compile every package because /home is a symlink to /var/home. This is because it's immutable and everything under / and /usr is read-only. Looking at the `brew` script it's using `pwd -P`; is there a reason the physical path is being used?

**Comment by SMillerDev** [comment]
My uneducated guess would be because applications you install with brew generally do the same thing.

@homebrew/Linux might know more though.

**Reply by iMichka** [reply]
The install script comes from the Mac world. It was then adapted for Linux. I guess no-one ever tried it on that Fedora version.

The script is at https://github.com/Homebrew/install and any improvement is welcome.

**Reply by sjackman** [reply]
The issue is not some much with the installer, but with the logic that determines whether a bottle can be used on this system. If `/home/linuxbrew/.linuxbrew/Cellar/` (or one of its components) is a symlink to `/var/home/linuxbrew/.linuxbrew/Cellar`, then Homebrew will determine that because the cellar is not at the default location, bottles cannot be used.

FWIW, this check can be defeated pretty easily defeated by modifying these lines:
https://github.com/Homebrew/brew/blob/bc9b98aa0ba777889218ac51f67ccc76b0be6a28/bin/brew#L31-L33

I had a work system where I did this, and it worked 99.9%. A few formulae fail. For example, the `postinstall` of `python` failed, which I fixed with a PR. There may be other examples.

**Reply by mdheller** [reply]
Has anyone made any progress on this? I'd be interested to see if anyones made any progress and any suggestions for approaching this (above the discussion above) are definitely appreciate--even if only conjecture : )

**Reply by jwhite007** [reply]
On RHEL 8.4, I have /home/linuxbrew as a symlink to /toolbox/linuxbrew. On my system, /toolbox is a mounted nfs share. In the most recent version of $(brew --prefix)/Homebrew/bin/brew, I’ve changed the “pwd -P” on both lines 19 and 31 to be simply “pwd”. For testing, I kept the brew script unchanged and made a copy, brew_mod, with the changes. Everything seems to be working very well even with bottle installs.

So far, I’ve installed python, pyenv, tmux, git, and gcc with this setup and all were installed from bottles. I also remember most if not all of the many respective dependencies having been installed from bottles as well. I've also installed several python packages. All of this, for the most part, went without glitches. I didn’t even get the python postinstall fail that @sjackman  speaks of at https://github.com/Linuxbrew/brew/issues/586. I did have a postinstall fail with gcc, but that is because I have yet to install the development tools to the system paths. The gcc postinstall debug showed that it was failing because /usr/bin/cc could not be found. I think it’s clear that the gcc postinstall fail has nothing to do with the symlinking. In addition, there are no symlinking-related warnings or errors that show in the output of ```brew dr```.

That is not to say that I'm not experiencing odd behavior, albeit inconsequential at this point.
```
$ diff <(brew_mod config) <(brew config)
9c9,11
< HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
---
> HOMEBREW_PREFIX: /toolbox/linuxbrew/.linuxbrew
> HOMEBREW_REPOSITORY: /toolbox/linuxbrew/.linuxbrew/Homebrew
> HOMEBREW_CELLAR: /toolbox/linuxbrew/.linuxbrew/Cellar
21,22c23,24
< Git: 2.33.0 => /home/linuxbrew/.linuxbrew/bin/git
< Curl: 7.78.0 => /home/linuxbrew/.linuxbrew/opt/curl/bin/curl
---
> Git: 2.33.0 => /toolbox/linuxbrew/.linuxbrew/bin/git
> Curl: 7.78.0 => /toolbox/linuxbrew/.linuxbrew/opt/curl/bin/curl
```
You'll notice in the above output that neither HOMEBREW_REPOSITORY nor HOMEBREW_CELLAR is being reported by ```brew_mod config```; however, you’ll see in the following that brew_mod does indeed know what they are.
 ```
$ brew_mod --repo
/home/linuxbrew/.linuxbrew/Homebrew
$ brew_mod --cellar
/home/linuxbrew/.linuxbrew/Cellar
 ```
Now I’m not claiming that this is a complete fix. It seems to be working very well for me so far; however, if it were really this simple, I'm sure that one of the maintainers would have already suggested it. Maybe a maintainer can comment as to what the complete consequences, from a Linuxbrew perspective, are of changing "pwd -P" to "pwd" on those two lines.

**Reply by jwhite007** [reply]
Update: I've discovered now that my solution above, while fixing the problem of getting bottles to install with a symlinked /home/linuxbrew, breaks the functionality of builds. My guess is that what @SMillerDev mentions above may allude to why this is so. Changing "pwd -P" to simply "pwd" results in every ```make install``` throwing ```/toolbox/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/make: line 30: /toolbox/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/make: Argument list too long``` Of course, one could switch back and forth between different forms of the brew script; however, if there are any dependencies which require a build when a bottle is being installed, those builds will fail.

**Reply by marcfargas** [reply]
See  https://github.com/Homebrew/install/issues/668 for more info on the issue. Does not refer to Silverblue but the underlying problem is the same. There is also https://github.com/Homebrew/brew/issues/14557

**Reply by osalbahr** [reply]
I'm not sure I understand the issue. Why would the argument list be too long when using "pwd" in lieu of "pwd -P"?

**Comment by ImBearChild** [comment]
I wrote a small tool for this. It will use bwrap to create a mount namespace without `/var/home` but `/home`. So brew inside this namespace can use bottles normally.

https://github.com/ImBearChild/brewenv.py

**Reply by osalbahr** [reply]
Nice! What are the advantages as opposed to `distrobox enter`?

**Reply by ImBearChild** [reply]
It's much smaller and lightweight. Distrobox is a conatainer with special configuration, basically. So it uses normal OCI container image, and user can not modify its configuration once after created (due to limits of OCI runtime). Brewenv is a small python script uses bwrap, and do not need extra image to work. And user can modify this script to fit their need (for example: use different mount point).

**Comment by stove-panini** [comment]
You can work around this by replacing the symlink with a bind mount. This should be done within a toolbox environment! That's the Silverblue way.

Add something like this to your `~/.bashrc`:
```sh
if [[ $TOOLBOX_PATH && -L /home ]]; then
    sudo rm /home
    sudo mkdir /home
    sudo mount -o bind /var/home /home
fi
```

Edited to be more concise, but I like to have an overarching `if [[ $TOOLBOX_PATH ]]; then...` to have toolbox-specific aliases & functions.

**Comment by tidux** [comment]
The whole reason to run Homebrew on Silverblue or Kinoite in the first place is to get tools that exist outside of both rpm-ostree and containers.  If you're going to play within those sandboxes then just use RPMs instead and deal with the reboots for system tools.

**Comment by mdheller** [comment]
For me, it's an ecosystem thing. I also don't like the speed / response
times when I update OSTree, so I believe a two kernel system is best. I'm
hoping to use a cross platform package management system like brew or maybe
nix as a means to bridge that gap, but Im still tinkering.

**Reply by osalbahr** [reply]
Is there still interest in using `brew` on Silverblue?

**Comment by akvadrako** [comment]
Why should homebrew break when /home is a symlink? To me this looks like a bug.

**Reply by osalbahr** [reply]
Yeah, not sure why a symlink would be an issue. I’ll see if I can figure out a way to fix it 

update: it’s more complicated than I thought

**Comment by osalbahr** [comment]
I just tried `brew doctor` and this is what I got. Not sure if it's intentional for it to fail on a symlink

```console
[zeus@fedora ~]$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /home/linuxbrew/.linuxbrew.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.

[zeus@fedora ~]$ brew --prefix
/var/home/linuxbrew/.linuxbrew
[zeus@fedora ~]$ realpath /home/linuxbrew/.linuxbrew
/var/home/linuxbrew/.linuxbrew
[zeus@fedora ~]$ 
```

I am currently using uBlue:
```console
[zeus@fedora ~]$ brew config
HOMEBREW_VERSION: 4.0.24
ORIGIN: https://github.com/Homebrew/brew
HEAD: 54c8876dc39047c04de15e7d212979ae8d98cf1c
Last commit: 7 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3d9284ae5a17aa2ed4500120cfd2ad8204035923
Core tap last commit: 20 hours ago
Core tap branch: master
Core tap JSON: 25 Jun 14:02 UTC
HOMEBREW_PREFIX: /var/home/linuxbrew/.linuxbrew
HOMEBREW_REPOSITORY: /var/home/linuxbrew/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /var/home/linuxbrew/.linuxbrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_INSTALL_FROM_API: set
Homebrew Ruby: 2.6.10 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit skylake
Clang: N/A
Git: 2.41.0 => /bin/git
Curl: 8.0.1 => /bin/curl
Kernel: Linux 6.3.8-200.fc38.x86_64 x86_64 GNU/Linux
OS: Fedora release 38 (Thirty Eight)
Host glibc: 2.37
/usr/bin/gcc: 13.1.1
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
[zeus@fedora ~]$ rpm-ostree status
...
● ostree-unverified-registry:ghcr.io/ublue-os/silverblue-nvidia:latest
                   Digest: sha256:ef7bc1f6435e33c458b3b109bf94805b0d12353d63b94af1f7756fcbecf0c476
                  Version: 38.20230625.0 (2023-06-25T13:00:02Z)
```

**Reply by osalbahr** [reply]
It looks like the check happens in [check_homebrew_prefix](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/diagnostic.rb#L808):
```console
[zeus@fedora ~]$ brew doctor --debug
==> Checking check_access_directories
==> Checking check_brew_git_origin
/usr/bin/env /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/shared/git --version
==> Checking check_casktap_integrity
==> Checking check_coretap_integrity
==> Checking check_deleted_formula
==> Checking check_deprecated_disabled
==> Checking check_deprecated_official_taps
==> Checking check_exist_directories
==> Checking check_for_external_cmd_name_conflict
==> Checking check_for_git
==> Checking check_for_installed_developer_tools
==> Checking check_for_non_prefixed_coreutils
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/coreutils.rb
==> Checking check_for_other_frameworks
==> Checking check_for_pydistutils_cfg_in_home
==> Checking check_for_stray_dylibs
==> Checking check_for_stray_headers
==> Checking check_for_stray_las
==> Checking check_for_stray_pcs
==> Checking check_for_stray_static_libs
==> Checking check_for_symlinked_cellar
==> Checking check_for_tap_ruby_files_locations
==> Checking check_for_unlinked_but_not_keg_only
==> Checking check_for_unnecessary_cask_tap
==> Checking check_for_unnecessary_core_tap
==> Checking check_for_unreadable_installed_formula
==> Checking check_gcc_dependent_linkage
==> Checking check_git_newline_settings
==> Checking check_git_status
==> Checking check_git_version
==> Checking check_glibc_minimum_version
==> Checking check_homebrew_prefix
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /home/linuxbrew/.linuxbrew.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.

==> Checking check_kernel_minimum_version
==> Checking check_linuxbrew_bottle_domain
==> Checking check_linuxbrew_core
==> Checking check_multiple_cellars
==> Checking check_supported_architecture
==> Checking check_tap_git_branch
==> Checking check_tmpdir
==> Checking check_tmpdir_executable
==> Checking check_tmpdir_sticky_bit
==> Checking check_umask_not_zero
==> Checking check_user_path_1
==> Checking check_user_path_2
==> Checking check_user_path_3
==> Checking check_xdg_data_dirs
==> Checking check_for_broken_symlinks
==> Checking check_missing_deps
[zeus@fedora ~]$ 
```

**Reply by osalbahr** [reply]
Would it be a good idea to check if `/var/home/linuxbrew/.linuxbrew` is the same as `/home/linuxbrew/.linuxbrew` and not fail?

Also, I am not sure how to change `HOMEBREW_PREFIX`:
```console
[zeus@fedora ~]$ brew config | grep PREFIX
HOMEBREW_PREFIX: /var/home/linuxbrew/.linuxbrew
[zeus@fedora ~]$ export HOMEBREW_PREFIX='/home/linuxbrew/.linuxbrew'
[zeus@fedora ~]$ brew config | grep PREFIX
HOMEBREW_PREFIX: /var/home/linuxbrew/.linuxbrew
```

**Reply by stove-panini** [reply]
The code that sets those variables is referenced in this comment from earlier in the thread: https://github.com/orgs/Homebrew/discussions/1282#discussioncomment-611816

`readlink` is used to resolve the real path of the home directory containing the `brew` executable. I documented a way around this using a bind mount instead, though it requires that brew is installed in your toolbox container: https://github.com/orgs/Homebrew/discussions/1282#discussioncomment-5067527

**Reply by osalbahr** [reply]
Thank you for the links. Yeah, it looks like these variables reset each time. No sure why they have to use the physical (real) path.

I saw your workaround. But I don't really get why `brew` shouldn't work outside of a toolbox.

**Comment by mdheller** [comment]
bigtime!

**Reply by osalbahr** [reply]
FYI, I was able to `brew install gcc` and then export `gcc-13` and it works fine without a bind mount or other workarounds. The only disadvantage is that you still have to individually export all you need. This is useless if your goal is to avoid using containers, though.

![image](https://github.com/Homebrew/discussions/assets/54853250/99cb97e2-78a5-4cb1-b467-11e819fbe9b1)

**Comment by osalbahr** [comment]
Thoughts on adding a warning to the installer that links to this discussion if /home is a symlink? 

Edit: opened Homebrew/install#786

**Comment by osalbahr** [comment]
Related issue at uBlue: [Ship homebrew config out of the box #252](https://github.com/ublue-os/bluefin/issues/252)

**Comment by osalbahr** [comment]
I think me and @rohanssrao found a patch to get Homebrew to work on Silverblue. We tested on containers where we symlinked /home to var/home. You can apply it as follows:
```console
sed -i "s/&& pwd -P)/\&\& pwd)/g" /home/linuxbrew/.linuxbrew/Homebrew/bin/brew
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
brew developer on
brew install xz
```
See https://github.com/ublue-os/bluefin/issues/252#issuecomment-1626646275.

You should then be able to install things normally. Feel free to test it and lmk how it goes!

The reproducible script that you can run in a container is at https://github.com/osalbahr/brew/blob/silverbrew-develop/silverbrew-fedora-toolbox.sh. It was tested on `fedora-toolbox` and `ubuntu`. I also tested it locally.

Edit: Well, to clarify the removal of `-P` was already mentioned. We only created and tested a proof-of-concept really.

**Reply by rohanssrao** [reply]
As far as we can tell, `brew developer on` is only needed now because it prevents brew from reverting the sed command by resetting the working directory to mirror master. If the script patch gets merged to Homebrew master, the developer command won't be needed.
In summary, the required change is the removal of two characters (`-P`) from bin/brew.

**Reply by osalbahr** [reply]
It seems that it doesn't cause issues on macOS:
```diff
~ % cd $(brew --prefix)/Homebrew
Homebrew % git diff
diff --git a/bin/brew b/bin/brew
index afc02e9eb..373294806 100755
--- a/bin/brew
+++ b/bin/brew
@@ -40,7 +40,7 @@ do
 done
 unset cmd
 
-BREW_FILE_DIRECTORY="$(quiet_cd "${0%/*}/" && pwd -P)"
+BREW_FILE_DIRECTORY="$(quiet_cd "${0%/*}/" && pwd)"
 HOMEBREW_BREW_FILE="${BREW_FILE_DIRECTORY%/}/${0##*/}"
 HOMEBREW_PREFIX="${HOMEBREW_BREW_FILE%/*/*}"
 
Homebrew % brew reinstall xz
==> Fetching xz
==> Downloading https://ghcr.io/v2/homebrew/core/xz/manifests/5.4.3
Already downloaded: /Users/dracula/Library/Caches/Homebrew/downloads/8d9f9cb1c34807246a161f74e3de29be172826599700e583c128fa2340fe177f--xz-5.4.3.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/xz/blobs/sha256:b9aebe3c04f5f17
Already downloaded: /Users/dracula/Library/Caches/Homebrew/downloads/b3f96a1454420c8df1d66c3fa5190d11a0dad2e939800646f188872176751c3b--xz--5.4.3.ventura.bottle.tar.gz
==> Reinstalling xz 
==> Pouring xz--5.4.3.ventura.bottle.tar.gz
🍺  /usr/local/Cellar/xz/5.4.3: 162 files, 2.5MB
==> Running `brew cleanup xz`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

**Comment by rohanssrao** [comment]
@osalbahr and I made a PR on the Homebrew repo with our fix: https://github.com/Homebrew/brew/pull/15656
It currently has 1 approval and we're waiting for it to be merged.

EDIT: It has been merged!

**Reply by osalbahr** [reply]
I think this should be selected as the answer.
